pkg/xen-tools/patches-4.15.0/14-qemu-Init-CPU-mask-per-VCPU.patch

```diff
@@ -0,0 +1,188 @@
+index b06f13e..a4b4f11 100644
+--- a/tools/qemu-xen/include/hw/boards.h
++++ b/tools/qemu-xen/include/hw/boards.h
+@@ -296,6 +296,7 @@ struct MachineState {
+     char *initrd_filename;
+     bool cpu_pin;
+     char *cpumask_str;
++    uint64_t cpumask;
+     const char *cpu_type;
+     AccelState *accelerator;
+     CPUArchIdList *possible_cpus;
+diff --git a/tools/qemu-xen/include/hw/core/cpu.h b/tools/qemu-xen/include/hw/core/cpu.h
+index 8f14573..628b31a 100644
+--- a/tools/qemu-xen/include/hw/core/cpu.h
++++ b/tools/qemu-xen/include/hw/core/cpu.h
+@@ -374,6 +374,8 @@ struct CPUState {
+     bool created;
+     bool stop;
+     bool stopped;
++    bool pinned;
++    uint64_t cpumask;
+     bool unplug;
+     bool crash_occurred;
+     bool exit_request;
+diff --git a/tools/qemu-xen/softmmu/cpus.c b/tools/qemu-xen/softmmu/cpus.c
+index a802e89..da56052 100644
+--- a/tools/qemu-xen/softmmu/cpus.c
++++ b/tools/qemu-xen/softmmu/cpus.c
+@@ -2011,6 +2011,23 @@ static void qemu_dummy_start_vcpu(CPUState *cpu)
+                        QEMU_THREAD_JOINABLE);
+ }
+ 
++static inline void cpumask_clear_bit(uint64_t *mask, uint8_t bit)
++{
++    *mask &= ~(1ul << bit);
++}
++
++static inline long cpumask_get_min_bit(uint64_t mask)
++{
++  return __builtin_ffsll(mask) - 1;
++}
++
++static long pick_pcpu(uint64_t *cpumask)
++{
++    long ret = cpumask_get_min_bit(*cpumask);
++    cpumask_clear_bit(cpumask, ret);
++    return ret;
++}
++
+ void qemu_init_vcpu(CPUState *cpu)
+ {
+     MachineState *ms = MACHINE(qdev_get_machine());
+@@ -2019,6 +2036,20 @@ void qemu_init_vcpu(CPUState *cpu)
+     cpu->nr_threads =  ms->smp.threads;
+     cpu->stopped = true;
+     cpu->random_seed = qemu_guest_random_seed_thread_part1();
++    cpu->pinned = ms->cpu_pin;
++    static uint64_t vm_cpumask;
++    uint64_t vcpu_cpumask;
++    if (!vm_cpumask)
++        vm_cpumask = ms->cpumask;
++    if (!cpu->pinned) {
++        /* If the CPUs are not pinned, assign the whole CPU mask to the VCPU */
++        vcpu_cpumask = vm_cpumask;
++    } else {
++        /* If the CPUs are pinned, pick only one CPU for this VCPU */
++        vcpu_cpumask = 1ull << pick_pcpu(&vm_cpumask);
++    }
++
++    cpu->cpumask = vcpu_cpumask;
+ 
+     if (!cpu->as) {
+         /* If the target cpu hasn't set up any address spaces itself,
+diff --git a/tools/qemu-xen/softmmu/vl.c b/tools/qemu-xen/softmmu/vl.c
+index c7c8abc..c7288e0 100644
+--- a/tools/qemu-xen/softmmu/vl.c
++++ b/tools/qemu-xen/softmmu/vl.c
+@@ -2829,6 +2829,72 @@ static void create_default_memdev(MachineState *ms, const char *path)
+                             &error_fatal);
+ }
+ 
++static inline unsigned get_max_cpu_in_mask(unsigned int cpumask)
++{
++    return (sizeof(cpumask) * BITS_PER_BYTE) - __builtin_clz(cpumask) - 1;
++}
++
++static inline void cpumask_set_bit(uint64_t *mask, uint8_t bit)
++{
++    *mask |= 1ull << bit ;
++}
++
++//Parse d[[,-]d]* mask (0-2 or 0-2,5,6). CPU starts with 0.
```

Comment:
I can see different description here and in .hx file: `CPU mask represented as a dec number. E.g. CPU mask 9 stands for 0b1001 and represents CPUs 0 and 3. It means that only physical CPUs 0 and 3 are available for the VM`. Should we adjust .hx, or am I missing something?

Reply:
Omg, I had one in mind but described another.
I have to fix the `.hx` file.
Thanks!

Reply:
Probably it would be great to add few word about options (with small samples) in commit message/PR description.

Reply:
Ok, will do

Reply:
done